neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -379,54 +388,272 @@ func QmpDelMemoryDevice(mon *qmp.SocketMonitor, idx int) error {
+		}
+		// Found unused backend to plug into
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpAddMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleUp",
+				fmt.Sprintf("Failed to add dimm%d to VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added dimm%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = true
+		r.memDevCount++
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveDevices() bool {
+	attempts := NewAttemptsCounter(r.memDevCount - r.targetCnt)
+	// Removing from the end to keep memslot1,memslot2,...
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || !active {
+			continue
+		}
+		// Found used backend to remove
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove dimm%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed dimm%d from VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.memDevCount--
 	}
-	var slot int
-	for slot = 1; slot <= int(slots); slot++ {
-		_, ok := backends[slot]
-		if !ok {
-			// that mean such object wasn't found, or by other words slot is empty
+
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveBackends() bool {
+	attempts := NewAttemptsCounter(len(r.memBackends) - r.targetCnt)
+
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+
+		if !attempts.Attempt() {
 			break
 		}
+
+		err := QmpDelMemoryBackend(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove memslot%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed memslot%d from VM %s", idx, r.vm.Name))
+		delete(r.memBackends, idx)
 	}
+	return attempts.DidSomething()
+}
 
-	err = QmpAddMemoryBackend(mon, slot, virtualmachine.Spec.Guest.MemorySlotSize.Value())
+func (r *QmpMemorySetter) run() (int, error) {
+	err := r.BuildState()
 	if err != nil {
-		return err
+		return -1, err
+	}
+
+	// Usually, runs first two or last two phases.
+	// If there are leftover slots, might run 2 and 4.
+	// If there are errors, last two phases serve as cleanup.
+	phases := []func() bool{
+		r.AddBackends,
+		r.AddDevices,
+		r.RemoveDevices,
+		r.RemoveBackends,
 	}
-	// now add pc-dimm device
-	err = QmpAddMemoryDevice(mon, slot)
+	for _, phase := range phases {
+		needSleep := phase()
+		if needSleep {
+			// This allows for QEMU and guest kernel to catch up.
+			time.Sleep(time.Second)
+		}
```

Comment:
Hm, I think "sleep just in case" is probably a bad idea... there's a limited number of reconcile worker threads, so if they start taking too long, the controller will start lagging on reconciling when we have many VMs.

Reply:
Ok, makes sense. I left it only in two phases, so it will sleep only once, as it were before.